timdex_dataset_api/utils.py

```diff
@@ -106,6 +108,109 @@ def _split_s3_uri(s3_uri: str) -> tuple[str, str]:
         return bucket, key
 
 
+class DuckDBConnectionFactory:
+    """Factory for creating and configuring DuckDB connections.
+
+    Args:
+        location_scheme: "file" or "s3", determines S3 credential setup
+        memory_limit: DuckDB memory limit (env: TDA_DUCKDB_MEMORY_LIMIT, default: "4GB")
+        threads: DuckDB thread limit (env: TDA_DUCKDB_THREADS, default: 8)
+    """
+
+    def __init__(
+        self,
+        location_scheme: Literal["file", "s3"] = "file",
+        memory_limit: str | None = None,
+        threads: int | None = None,
+    ):
+        self.location_scheme = location_scheme
+        self.memory_limit = memory_limit or os.getenv("TDA_DUCKDB_MEMORY_LIMIT", "4GB")
+        self.threads = threads or int(os.getenv("TDA_DUCKDB_THREADS", "8"))
+
+    def create_connection(self, path: str = ":memory:") -> DuckDBPyConnection:
+        """Create a new configured DuckDB connection.
+
+        Args:
+            path: Database file path or ":memory:" for in-memory database (default)
+        """
+        start_time = time.perf_counter()
+        conn = duckdb.connect(path)
+        conn.execute("SET enable_progress_bar = false;")
+        self.configure_connection(conn)
+        logger.debug(
+            f"DuckDB connection created, {round(time.perf_counter()-start_time,2)}s"
+        )
+        return conn
```

Comment:
This is the only way a DuckDB connection should be created now, from anywhere and for any reason.